javascript/ql/src/semmle/javascript/dataflow/Configuration.qll

```diff
@@ -483,6 +484,72 @@ private predicate reachableFromStoreBase(string prop, DataFlow::Node rhs, DataFl
     newSummary.valuePreserving() = true and
     summary = oldSummary.append(newSummary)
   )
+  or
+  nestedPropFlow(prop, rhs, nd, cfg, summary)
+}
+
+/**
+ * Holds if `rhs` is the right-hand side of a write to property `outerProp` and some read of
+ * another property `innerProp` is reachable from the base of that write under configuration `cfg`,
+ * and from the base of that inner read we can reach a read `succ` of the same property `innerProp`,
+ * such that the path from `rhs` to `succ` is summarized by `summary`.
+ *
+ * Example:
+ *
+ * ```
+ * let root = new A();
+ * let base = root.innerProp;
+ * base.outerProp = rhs;
+ * let succ = root.innerProp;
+ * ```
+ */
+private predicate nestedPropFlow(string outerProp, DataFlow::Node rhs, DataFlow::Node succ,
+                                 DataFlow::Configuration cfg, PathSummary summary) {
+  exists (DataFlow::PropRead nestedRead, PathSummary oldSummary |
+    reachableFromStoreBase(outerProp, rhs, nestedRead, cfg, oldSummary) and
+    loadLoadPair(nestedRead, succ, cfg, oldSummary, summary)
+  )
+}
+
+/**
+ * Holds if `load` is a read of some property `innerProp` from which we can reach a read `succ` of the same
+ * property `innerProp` under configuration `cfg`, and the concatenation of `oldSummary` with the summary
+ * of that path is `summary`.
+ *
+ * Example:
+ *
+ * ```
+ * let root = A();
+ * let load = root.innerProp;
+ * let succ = root.innerProp;
+ * ```
+ */
+pragma[noinline]
+private predicate loadLoadPair(DataFlow::PropRead load, DataFlow::Node succ,
+                               DataFlow::Configuration cfg, PathSummary oldSummary, PathSummary summary) {
+  exists (string innerProp, DataFlow::Node nd, PathSummary newSummary |
+    reachableFromLoadBase(innerProp, load, nd, cfg, newSummary) and
+    loadStep(nd, succ, innerProp) and
+    summary = oldSummary.append(newSummary)
+  )
```

Comment:
I don't understand why we have two `PathSummary` arguments. `oldSummary` is almost not bound at all; it's just anything that can be prefixed onto `newSummary` to yield `summary`.

It seems like the summary of this edge is a pair of input/output summaries, instead of just being a summary. I mean that's what `PathSummary` is for, right?

Apart from that, is it correctly understood that we would get N^2 load-load pairs for code like this?
```
elm.style.color = X;
elm.style.backgroundColor = X;
elm.style.border = X;
elm.style.padding = X;
...
```